### PR TITLE
misc. move typing improvements, WIP

### PR DESCRIPTION
**Pull request by a-detiste**
Hi,

I'm back working on this. I don't mind doing the + `from __future__ import`      - `Union & Optional` by hand if pyupgrade can't.

https://github.com/duecredit/duecredit/pull/191

### CODE REVIEW

**Comment by a-detiste** [comment]
@jwodder @yarikoptic

**Review by jwodder — CHANGES_REQUESTED** [review]

File: duecredit/tests/test__main__.py [hunk 1]
```diff
@@ -40,8 +39,8 @@ def test_main_version(monkeypatch: 'MonkeyPatch') -> None:
 
 
 def test_main_run_a_script(
-    tmpdir: 'TempdirFactory',
-    monkeypatch: 'MonkeyPatch'
+    tmpdir: TempPathFactory,
```
[the remark this answers is not in the document]
Comment (a-detiste) [reply duecredit/tests/test__main__.py]: thanks

File: duecredit/collector.py [hunk 2]
```diff
@@ -7,23 +7,27 @@
...
 """Citation and citations Collector classes"""
+from __future__ import annotations
 
+import logging
 import os
 import sys
+from collections import namedtuple
 from functools import wraps
-from typing import Any, Dict, List, Optional, Tuple, Union
+from typing import Any, Dict, List, Optional, Tuple, Union, TYPE_CHECKING
```
[the remark this answers is not in the document]
Comment (a-detiste) [reply duecredit/collector.py]: I modified only 1 file this way to let CI run.

File: duecredit/cmdline/cmd_summary.py [hunk 3]
```diff
@@ -20,10 +22,13 @@
 
 __docformat__ = 'restructuredtext'
 
+if TYPE_CHECKING:
+    import argparse
```
Comment (jwodder) [inline duecredit/cmdline/cmd_summary.py:26]: There's no downside to importing `argparse`.  Why gate it behind `if TYPE_CHECKING`?

File: duecredit/cmdline/cmd_test.py [hunk 4]
```diff
@@ -7,22 +7,28 @@
...
-
+from __future__ import annotations
 
 __docformat__ = 'restructuredtext'
 
+from typing import TYPE_CHECKING
+
+
+if TYPE_CHECKING:
+    import argparse
```
Comment (jwodder) [inline duecredit/cmdline/cmd_test.py:18]: Again, no reason for `if TYPE_CHECKING`.

**Review by jwodder — CHANGES_REQUESTED** [review]

File: duecredit/entries.py [hunk 5]
```diff
@@ -7,23 +7,24 @@
...
 
-    def __init__(self, rawentry:str, key: Optional[str] = None) -> None:
+    def __init__(self, rawentry:str, key: str | None = None) -> None:
         self._rawentry = rawentry
         self._key = key or rawentry.lower()
 
     def __eq__(self, other: object) -> bool:
         if not isinstance(other, DueCreditEntry):
-            raise NotImplemented
+            raise NotImplementedError
```
Comment (jwodder) [inline duecredit/entries.py:27]:
```suggestion
            return NotImplemented
```

**Review by jwodder — CHANGES_REQUESTED** [review]

File: duecredit/entries.py [hunk 6]
```diff
@@ -7,23 +7,24 @@
...
 
-    def __init__(self, rawentry:str, key: Optional[str] = None) -> None:
+    def __init__(self, rawentry:str, key: str | None = None) -> None:
         self._rawentry = rawentry
         self._key = key or rawentry.lower()
 
     def __eq__(self, other: object) -> bool:
         if not isinstance(other, DueCreditEntry):
-            raise NotImplemented
+            return NotImplemented
```
Comment (yarikoptic) [inline duecredit/entries.py:27]: I don't think it makes sense to return here

Comment (jwodder) [reply duecredit/entries.py:27]: It does make sense: https://docs.python.org/3/library/constants.html#NotImplemented

Comment (yarikoptic) [reply duecredit/entries.py:27]: ah -- cool, learned something new again, should have checked.

<details>
<summary>FWIW - here is a trial on a simple case</summary> 

```shell
❯ python -m trace --trace notimpl.py
 --- modulename: notimpl, funcname: <module>
notimpl.py(1): class A:
 --- modulename: notimpl, funcname: A
notimpl.py(1): class A:
notimpl.py(2):     def __eq__(self, other):
notimpl.py(8): print(A() == A())
 --- modulename: notimpl, funcname: __eq__
notimpl.py(3):         if isinstance(other, A):
notimpl.py(4):             return True
True
notimpl.py(9): print(A() == 1)
 --- modulename: notimpl, funcname: __eq__
notimpl.py(3):         if isinstance(other, A):
notimpl.py(6):             return NotImplemented
False
notimpl.py(10): print(A() != 1)
 --- modulename: notimpl, funcname: __eq__
notimpl.py(3):         if isinstance(other, A):
notimpl.py(6):             return NotImplemented
True
❯ cat notimpl.py
class A:
    def __eq__(self, other):
        if isinstance(other, A):
            return True
        else:
            return NotImplemented

print(A() == A())
print(A() == 1)
print(A() != 1)
```
</details>

I will revert that f470c3e8c41a02f70f0a700d81e7f1b2533abb60 I merged in rush, directly in master

**Pull request merged**